src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriter.cs

```diff
@@ -0,0 +1,202 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers.Binary;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Data.SqlClientX.IO
+{
+    /// <summary>
+    /// This class provides helper methods for writing bytes using <see cref="TdsStream"/>
+    /// It extends <see cref="TdsBufferAlloc"/> that manages allocations of bytes buffer for better memory management.
+    /// </summary>
+    internal class TdsWriter : IAsyncDisposable, IDisposable
+    {
+        private readonly TdsBufferAlloc _alloc;
+
+        /// <summary>
+        /// Instantiate TdsWriter with a TdsWriteStream
+        /// </summary>
+        /// <exception cref="ArgumentNullException"></exception>
+        public TdsWriter()
+        {
+            _alloc = new TdsBufferAlloc();
+        }
+
+        #region Public APIs
+
+        /// <summary>
+        /// Writes short value to out buffer, as little-endian.
+        /// </summary>
+        /// <param name="stream">Tds Stream to work with</param>
+        /// <param name="v">Value to write</param>
+        /// <param name="isAsync">Whether caller method is executing asynchronously.</param>
+        /// <param name="ct">Cancellation token</param>
+        public async ValueTask WriteShortAsync(TdsStream stream, short v, bool isAsync, CancellationToken ct)
+        {
+            var len = sizeof(short);
+            BinaryPrimitives.WriteInt16LittleEndian(_alloc.GetBuffer(len), v);
+            await WriteBytesAsync(stream, _alloc.GetBuffer(len), 0, len, isAsync, ct).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Writes int value to out buffer, as little-endian.
+        /// </summary>
+        /// <param name="stream">Tds Stream to work with</param>
+        /// <param name="v">Value to write</param>
+        /// <param name="isAsync">Whether caller method is executing asynchronously.</param>
+        /// <param name="ct">Cancellation token</param>
+        public async ValueTask WriteIntAsync(TdsStream stream, int v, bool isAsync, CancellationToken ct)
+        {
+            var len = sizeof(int);
+            BinaryPrimitives.WriteInt32LittleEndian(_alloc.GetBuffer(len), v);
+            await WriteBytesAsync(stream, _alloc.GetBuffer(len), 0, len, isAsync, ct).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Writes unsigned short value to out buffer, as little-endian.
+        /// </summary>
+        /// <param name="stream">Tds Stream to work with</param>
+        /// <param name="v">Value to write</param>
+        /// <param name="isAsync">Whether caller method is executing asynchronously.</param>
+        /// <param name="ct">Cancellation token</param>
+        public ValueTask WriteUnsignedShortAsync(TdsStream stream, ushort v, bool isAsync, CancellationToken ct)
```

Comment:
This feels excessive for a method name 😅 The pattern for classes like BitConverter and BinaryPrimitives is eg Uint16 or Int32. It also removes ambiguity for the size of the value to write - but I think dotnet defines the sizes of types like `short` , `long`, etc, in bits anyways (rather than words like C++ might).

Reply:
I think the existing method names would allow an easier migration from TdsParser / TdsParserStateobject callers to TdsWriter. 
I am hoping we can retain names which looks same as older analogous names in existing client.